File: Week1/Design_Principles_and_Patterns/Exercise1_Singleton_Pattern/Logger.java
// Logger.java
public class Logger {
    // Step 2: Private static instance
    private static Logger instance;

    // Step 2: Private constructor
    private Logger() {
        System.out.println("Logger instance created.");
    }

    // Step 2: Public static method to get the instance
    public static Logger getInstance() {
        if (instance == null) {
            instance = new Logger();  // Lazy initialization
        }
        return instance;
    }

    // Logging method
    public void log(String message) {
        System.out.println("[LOG]: " + message);
    }
}

File: Week1/Design_Principles_and_Patterns/Exe1_Singleton_Pattern/Logger.cs
using System;

public class Logger
{
    // Private static instance of Logger
    private static Logger? instance = null;

    // Object for locking to make thread-safe
    private static readonly object padlock = new object();

    // Private constructor to prevent external instantiation
    private Logger()
    {
        Console.WriteLine("Logger instance created.");
    }

    // Public static method to get the singleton instance
    public static Logger GetInstance()
    {
        if (instance == null)
        {
            lock (padlock)
            {
                if (instance == null)
                {
                    instance = new Logger();
                }
            }
        }
        return instance;
    }

    // Example logging method
    public void Log(string message)
    {
        Console.WriteLine($"[LOG] {message}");
    }
}
